### PR TITLE
[Cisco ASA]: Add destination.address to events with message ID 7250xx

### LINKED ISSUE #11290: [CISCO ASA]: Events with Cod 7250** could have destination.address insteat of destination.ip
### Integration Name

Cisco ASA [cisco_asa]

### Dataset Name

cisco_asa.log

### Integration Version

2.37.0

### Agent Version

8.15.0

### Agent Output Type

elasticsearch

### Elasticsearch Version

8.15.1

### OS Version and Architecture

RHEL 8

### Software/API Version

_No response_

### Error Message

Processor geoip with tag  in pipeline logs-cisco_asa.log-2.37.0 failed with message: 'hostnameabc.exmaple.com' is not an IP string literal.

### Event Original

<166>Oct 01 2024 14:03:24 ASAFW100 : %ASA-6-725001: Starting SSL handshake with client outside:IPV6ADDRESS to hostnameabc.example.com/443 for DTLS session

### What did you do?

Standard Configuration

### What did you see?

pipeline errors

### What did you expect to see?

A ECS conform document with all relevant fields which are expected after a complete pipeline processing

###  State
With the ASA message codes 725001,725002,725007 and 725016 it is possible that a destination address will be written instead of a destination IP address. In the corresponding grok processors, however, only ip addresses are taken into account or only the destination.ip field is created

As a result, a dfqdn is written in the destination.ip field. The subsequent geo-ip processors then fail. The processors for the geo-Ip do not ignore errors, which prevents further processing.

### Possible Solution
the only logical way would be to take this into account in the grok patterns instead of using %{NOTSPACE:destination.ip}. And possibly also introduce "ignore failures" in geoprocessing.

Affected lines:
999-1023:
``` - grok:
      field: message
      if: "ctx._temp_.cisco.message_id == '725001'"
      tag: parse_725001
      patterns:
        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
  - grok:
      field: message
      if: "ctx._temp_.cisco.message_id == '725002'"
      tag: parse_725002
      patterns:
        - '^Device completed SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version}'
  - grok:
      field: message
      if: "ctx._temp_.cisco.message_id == '725007'"
      tag: parse_725007
      patterns:
        - '^SSL session with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} terminated'
  - grok:
      field: message
      if: "ctx._temp_.cisco.message_id == '725016'"
      tag: parse_725016
      patterns:
        - '^Device selects trust-point %{DATA:_temp_.cisco.trustpoint} for %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port}$'
  ```     

### Remark

Perhaps it would be generally useful to consider this in the entire pipeline for the IP's in source and destination? To avoid errors. %{NOTSPACE} is not really smart in most cases, is it?

### PR DESCRIPTION
**Pull request by xtruthx**
## Proposed commit message

```
Introduce destination.address to message IDs 725001, 725002, 725007, and 725016.

Closes #11290
```

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] 

## Related issues
Closes #11290

### CODE REVIEW

File: packages/cisco_asa/changelog.yml [hunk 1]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "2.37.1"
+  changes:
+    - description: "Introduce destination address to 725001,725002,725007 and 725016"
```
Comment (andrewkroh) [inline packages/cisco_asa/changelog.yml:4]:
```suggestion
    - description: Introduce destination.address to message ID 725001, 725002, 725007, and 725016.
```

File: packages/cisco_asa/changelog.yml [hunk 2]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "2.37.1"
```
Comment (andrewkroh) [inline packages/cisco_asa/changelog.yml:2]: Under https://semver.org/ this version should be 2.38.0.

File: packages/cisco_asa/changelog.yml [hunk 3]
```diff
@@ -1,4 +1,9 @@
...
+  changes:
+    - description: "Introduce destination address to 725001,725002,725007 and 725016"
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/10897
```
Comment (andrewkroh) [inline packages/cisco_asa/changelog.yml:6]:
```suggestion
      link: https://github.com/elastic/integrations/pull/11292
```

File: packages/cisco_asa/changelog.yml (hunk 1 above)
Comment (xtruthx) [reply packages/cisco_asa/changelog.yml]: done

File: packages/cisco_asa/manifest.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 format_version: "3.0.3"
 name: cisco_asa
 title: Cisco ASA
-version: "2.37.0"
+version: "2.38.0
```
Comment (andrewkroh) [inline packages/cisco_asa/manifest.yml:4]:
```suggestion
version: "2.38.0"
```

File: packages/cisco_asa/changelog.yml (hunk 3 above)
Comment (xtruthx) [reply packages/cisco_asa/changelog.yml]: i forgot do adapt the pr number. done

File: packages/cisco_asa/changelog.yml (hunk 2 above)
Comment (xtruthx) [reply packages/cisco_asa/changelog.yml]: done

File: packages/cisco_asa/manifest.yml (hunk 4 above)
Comment (xtruthx) [reply packages/cisco_asa/manifest.yml]: thx

**Review by nicholasberlin — APPROVED** [review]

**Review by taylor-swanson — COMMENTED** [review]
Could you add pipeline test cases to show the new patterns work?

The [test-asa.log](https://github.com/elastic/integrations/blob/main/packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-asa.log) file should be fine.

As a note, `destination.address` can hold either an IP or hostname. `destination.domain` is a better fit for a hostname-only value.

https://www.elastic.co/guide/en/ecs/current/ecs-destination.html

As the note mentions for `destination.address`, the value should duplicated to either `destination.ip` or `destination.domain` depending on what type it is. That would be a more complicated change and we'd also have to consider interactions with other message IDs, so I wouldn't worry about it for this change.

**Comment by xtruthx** [comment]
you're right. destination.domain is correct, I made a mistake. I would say simply checking whether IP or domain via the GROK pattern should be enough. Or not?

I will add the tests with changes.

**Comment by taylor-swanson** [comment]
> I would say simply checking whether IP or domain via the GROK pattern should be enough. Or not?

I think both approaches have merit.

If we rely on GROK, then we just copy the values to `destination.address`. The end result is the same, we just work backwards compared to extracting to `destination.address` first, then rely on a convert processor to detect if it's an IP or not. The downside with GROK is that it won't be immediately obvious if parsing fails due to a bad ip/hostname value. I don't have strong enough opinions either way to say which way to go. I'm fine with keeping the GROK as is for now.

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -1001,26 +1001,26 @@ processors:
       if: "ctx._temp_.cisco.message_id == '725001'"
       tag: parse_725001
       patterns:
-        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
+        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to (?:%{IP:destination.ip}|%{HOSTNAME:destination.address})?/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
```
Comment (taylor-swanson) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1004]:
```suggestion
        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.address}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.address}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
```

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 6]
```diff
@@ -1001,26 +1001,26 @@ processors:
...
         - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
```
Comment (taylor-swanson) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1005]:
```suggestion
        - '^Starting SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.address}/%{NOTSPACE:source.port} for %{NOTSPACE:_temp_.cisco.tls_version} session'
```

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -1001,26 +1001,26 @@ processors:
...
   - grok:
       field: message
       if: "ctx._temp_.cisco.message_id == '725002'"
       tag: parse_725002
       patterns:
-        - '^Device completed SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version}'
+        - '^Device completed SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to (?:%{IP:destination.ip}|%{HOSTNAME:destination.address})?/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version}'
```
Comment (taylor-swanson) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1011]:
```suggestion
        - '^Device completed SSL handshake with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.address}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.address}/%{NOTSPACE:destination.port} for %{NOTSPACE:_temp_.cisco.tls_version}'
```

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -1001,26 +1001,26 @@ processors:
...
   - grok:
       field: message
       if: "ctx._temp_.cisco.message_id == '725007'"
       tag: parse_725007
       patterns:
-        - '^SSL session with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port} terminated'
+        - '^SSL session with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to (?:%{IP:destination.ip}|%{HOSTNAME:destination.address})?/%{NOTSPACE:destination.port} terminated'
```
Comment (taylor-swanson) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1017]:
```suggestion
        - '^SSL session with %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.address}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.address}/%{NOTSPACE:destination.port} terminated'
```

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 9]
```diff
@@ -1001,26 +1001,26 @@ processors:
...
   - grok:
       field: message
       if: "ctx._temp_.cisco.message_id == '725016'"
       tag: parse_725016
       patterns:
-        - '^Device selects trust-point %{DATA:_temp_.cisco.trustpoint} for %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.ip}/%{NOTSPACE:destination.port}$'
+        - '^Device selects trust-point %{DATA:_temp_.cisco.trustpoint} for %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.ip}/%{NOTSPACE:source.port} to (?:%{IP:destination.ip}|%{HOSTNAME:destination.address})?/%{NOTSPACE:destination.port}$'
```
Comment (taylor-swanson) [inline packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1023]:
```suggestion
        - '^Device selects trust-point %{DATA:_temp_.cisco.trustpoint} for %{NOTSPACE:_temp_.cisco.peer_type} %{DATA:_temp_.cisco.source_interface}:%{NOTSPACE:source.address}/%{NOTSPACE:source.port} to %{NOTSPACE:destination.address}/%{NOTSPACE:destination.port}$'
```

**Review by taylor-swanson — CHANGES_REQUESTED** [review]
On closer inspection of the pipeline, there are a couple of groks towards the end that handle extracting the `*.address` values to either ip or domain. So for this changes, let's revert back to the old patterns, but change where they extract to to the `*.address` fields.

When updating the expected test files, existing tests should only show the addition of the `*.address` fields, they should still extract to the `*.ip` fields.

**Comment by xtruthx** [comment]
Ok, I haven't seen them. Then of course it makes more sense to fill *.address and then let the GROK processors decide. 

For the processors, however, I would not want to set the pattern for *.domain to GREEDYDATA but rather to HOSTNAME.
I would also prefer to use HOSTNAME

**Comment by xtruthx** [comment]
but I can't find an event in the tests that contains a domain?

**Comment by taylor-swanson** [comment]
> For the processors, however, I would not want to set the pattern for *.domain to GREEDYDATA but rather to HOSTNAME. I would also prefer to use HOSTNAME

No disagreement there. I think the intent was for it to set *.ip if it's an IP, otherwise everything else goes to *.domain. That code has been there for over 3 years now and we haven't received any complaints on it yet as far as I know. 🤷

I triggered CI to run, but it will fail due to the expected files not being updated.

Could you run the following and commit the changes?

```
elastic-package test pipeline -g
```

This will update the pipeline test expected files so they will pass.

**Comment by xtruthx** [comment]
> No disagreement there. I think the intent was for it to set *.ip if it's an IP, otherwise everything else goes to *.domain. That code has been there for over 3 years now and we haven't received any complaints on it yet as far as I know.

I will do that.


It was not meant as a complaint but as an improvement. Wildcard patterns are more expensive than precise patterns

**Comment by xtruthx** [comment]
```
xtruthx   cisco-asa  …  integrations-fork  packages  cisco_asa  elastic-package test pipeline -g
Run pipeline tests for the package
--- Test results for package: cisco_asa - START ---
╭───────────┬─────────────┬───────────┬─────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM │ TEST TYPE │ TEST NAME                                               │ RESULT │ TIME ELAPSED │
├───────────┼─────────────┼───────────┼─────────────────────────────────────────────────────────┼────────┼──────────────┤
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-additional-messages.log) │ PASS   │ 134.795116ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-anyconnect-messages.log) │ PASS   │ 144.740119ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-asa-fix.log)             │ PASS   │ 137.902587ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-asa-missing-groups.log)  │ PASS   │ 136.258244ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-asa.log)                 │ PASS   │ 138.979295ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-dap-records.log)         │ PASS   │ 136.255288ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-filtered.log)            │ PASS   │   136.2838ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-hostnames.log)           │ PASS   │ 136.799084ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-non-canonical.log)       │ PASS   │ 130.839458ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-not-ip.log)              │ PASS   │ 152.314736ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-sample.log)              │ PASS   │ 168.144326ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-sgt-tag-name.log)        │ PASS   │ 143.442329ms │
│ cisco_asa │ log         │ pipeline  │ (ingest pipeline warnings test-sip.log)                 │ PASS   │    150.306ms │
│ cisco_asa │ log         │ pipeline  │ test-additional-messages.log                            │ PASS   │  729.07818ms │
│ cisco_asa │ log         │ pipeline  │ test-anyconnect-messages.log                            │ PASS   │ 144.167914ms │
│ cisco_asa │ log         │ pipeline  │ test-asa-fix.log                                        │ PASS   │ 164.375689ms │
│ cisco_asa │ log         │ pipeline  │ test-asa-missing-groups.log                             │ PASS   │ 147.380757ms │
│ cisco_asa │ log         │ pipeline  │ test-asa.log                                            │ PASS   │ 1.247635801s │
│ cisco_asa │ log         │ pipeline  │ test-dap-records.log                                    │ PASS   │  89.198875ms │
│ cisco_asa │ log         │ pipeline  │ test-filtered.log                                       │ PASS   │  92.931751ms │
│ cisco_asa │ log         │ pipeline  │ test-hostnames.log                                      │ PASS   │  98.451796ms │
│ cisco_asa │ log         │ pipeline  │ test-non-canonical.log                                  │ PASS   │ 175.226643ms │
│ cisco_asa │ log         │ pipeline  │ test-not-ip.log                                         │ PASS   │ 106.357273ms │
│ cisco_asa │ log         │ pipeline  │ test-sample.log                                         │ PASS   │ 542.359218ms │
│ cisco_asa │ log         │ pipeline  │ test-sgt-tag-name.log                                   │ PASS   │  236.40216ms │
│ cisco_asa │ log         │ pipeline  │ test-sip.log                                            │ PASS   │ 113.469469ms │
╰───────────┴─────────────┴───────────┴─────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_asa - END   ---
Done
```

**Review by taylor-swanson — APPROVED** [review]
LGTM!

**Pull request merged**